src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -1282,4 +1282,97 @@ BOOST_AUTO_TEST_CASE(SimpleMeanSquaredErrorLayerTest)
+            1 * j / (double)outColSize + 1 * d / (double)depth;
+      }
+
+  BilinearInterpolation<> layer(inRowSize, inColSize, outRowSize, outColSize,
+                                depth);
+
+  layer.Forward(std::move(input), std::move(output));
+  CheckMatrices(output - expectedOutput, arma::zeros(output.n_rows), 1e-12);
+
+  layer.Backward(std::move(output), std::move(output),
+        std::move(unzoomedOutput));
+  CheckMatrices(unzoomedOutput - input, arma::zeros(input.n_rows), 1e-12);
+
+  arma::vec x = arma::linspace<arma::vec>(1, 2, 2);
+  arma::vec y = arma::linspace<arma::vec>(2, 3, 2);
+  //std::cout << std::endl << "x" << std::endl << x << std::endl;
+  //std::cout << std::endl << "y" << std::endl << y << std::endl;
+  arma::vec ww = arma::linspace<arma::vec>(1, 2, 3);
+  arma::vec xx = arma::linspace<arma::vec>(2, 3, 3);
+  arma::vec yy, zz;
+  arma::vec j = ((4 * x) - 1) / 3;
+  arma::vec k = ((4 * y) - 1) / 3;
+  arma::interp1(x, j, ww, yy);
+  arma::interp1(y, k, xx, zz);
+  //std::cout << std::endl << "ww" << std::endl << ww << std::endl;
+  //std::cout << std::endl << "yy" << std::endl << yy << std::endl;
+  //std::cout << std::endl << "xx" << std::endl << xx << std::endl;
+  //std::cout << std::endl << "zz" << std::endl << zz << std::endl;
+  arma::rowvec x1({yy[0], zz[0]});
+  arma::rowvec x2({yy[1], zz[1]});
+  arma::rowvec x3({yy[2], zz[2]});
+  //std::cout << std::endl << "x1" << std::endl << x1 << std::endl;
+  //std::cout << std::endl << "x2" << std::endl << x2 << std::endl;
+  //std::cout << std::endl << "x3" << std::endl << x3 << std::endl;
+  arma::rowvec aa = arma::linspace<arma::rowvec>(x1[0], x1[1], 3);
+  arma::rowvec bb = arma::linspace<arma::rowvec>(x2[0], x2[1], 3);
+  arma::rowvec cc = arma::linspace<arma::rowvec>(x3[0], x3[1], 3);
+  //std::cout << std::endl << "aa" << std::endl << aa << std::endl;
+  //std::cout << std::endl << "bb" << std::endl << bb << std::endl;
+  //std::cout << std::endl << "cc" << std::endl << cc << std::endl;
+  input.zeros(2 * 2 * 1, 1);
+  input[0] = x[0];
+  input[1] = x[1];
+  input[2] = y[0];
+  input[3] = y[1];
+  //std::cout << std::endl << "input" << std::endl << input << std::endl;
+  output.zeros(3 * 3 * 1, 1);
+  BilinearInterpolation<> layer1(2, 2, 3, 3, 1);
+  layer1.Forward(std::move(input), std::move(output));
+  //std::cout << std::endl << "output1" << std::endl << output << std::endl;
+  expectedOutput.set_size(3 * 3 * 1, 1);
+  expectedOutput[0] = aa[0];
+  expectedOutput[1] = aa[1];
+  expectedOutput[2] = aa[2];
+  expectedOutput[3] = bb[0];
+  expectedOutput[4] = bb[1];
+  expectedOutput[5] = bb[2];
+  expectedOutput[6] = cc[0];
+  expectedOutput[7] = cc[1];
+  expectedOutput[8] = cc[2];
+  //std::cout << std::endl << "expectedOutput" << std::endl << expectedOutput << std::endl;
+  CheckMatrices(output - expectedOutput, arma::zeros(output.n_rows), 1e-12);
+  //std::cout << std::endl << "expectedOutput" << std::endl << expectedOutput << std::endl;
```

Comment:
@zoq I could only come up with this sort of hacky version, seemingly `arma::interp1` doesn't work the way bilinear-interpolation does.

Reply:
I see, do you think we should instead compare against a precomputed result?

Reply:
A precomputed result should be better. Probably from `Tensorflow` or `Skimage`? But we already have a test-case that tests that above.

Reply:
@zoq If you could suggest a source, it'd be great. I'd drop these changes then.

Reply:
"But we already have a test-case that tests that above." Do you mean the three for loops?

Reply:
I think comparing against skimage does make sense, this is a really simple test.

Reply:
Yes, those were a part of Kris' original test.

Reply:
I'll see what can be done.

Reply:
Would be great if you could compare it a precomputed result, skimage or tensorflow is just fine for me.

Reply:
This turned out to be a good idea @zoq. Found a bug in Kris' implementation logic that was going unnoticed.